src/graph.rs

```diff
@@ -130,6 +167,843 @@ impl Handle<Symbol> {
+        }
+    }
+}
+
+//-------------------------------------------------------------------------------------------------
+// Nodes
+
+/// Uniquely identifies a node in a stack graph.
+///
+/// Each node (except for the _root node_ and _jump to scope_ node) lives in a file, and has a
+/// _local ID_ that must be unique within its file.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct NodeID {
+    /// The file that the node comes from.
+    pub file: Handle<File>,
+    /// The unique identity of the node within its file.
+    pub local_id: u32,
+}
+
+#[doc(hidden)]
+pub struct DisplayNodeID<'a> {
+    wrapped: NodeID,
+    graph: &'a StackGraph,
+}
+
+impl<'a> Display for DisplayNodeID<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(
+            f,
+            "{}({})",
+            self.wrapped.file.display(self.graph),
+            self.wrapped.local_id,
+        )
+    }
+}
+
+impl NodeID {
+    pub fn display(self, graph: &StackGraph) -> impl Display + '_ {
+        DisplayNodeID {
+            wrapped: self,
+            graph,
+        }
+    }
+}
+
+/// A node in a stack graph.
+pub enum Node {
+    DropScopes(DropScopesNode),
+    ExportedScope(ExportedScopeNode),
+    InternalScope(InternalScopeNode),
+    JumpTo(JumpToNode),
+    PushScopedSymbol(PushScopedSymbolNode),
+    PushSymbol(PushSymbolNode),
+    PopScopedSymbol(PopScopedSymbolNode),
+    PopSymbol(PopSymbolNode),
+    Root(RootNode),
+    Unknown(UnknownNode),
+}
+
+impl Node {
+    #[inline(always)]
+    pub fn is_definition(&self) -> bool {
+        match self {
+            Node::PopScopedSymbol(node) => node.is_definition,
+            Node::PopSymbol(node) => node.is_definition,
+            _ => false,
+        }
+    }
+
+    #[inline(always)]
+    pub fn is_reference(&self) -> bool {
+        match self {
+            Node::PushScopedSymbol(node) => node.is_reference,
+            Node::PushSymbol(node) => node.is_reference,
+            _ => false,
+        }
+    }
+
+    #[inline(always)]
+    pub fn is_jump_to(&self) -> bool {
+        matches!(self, Node::JumpTo(_))
+    }
+
+    #[inline(always)]
+    pub fn is_root(&self) -> bool {
+        matches!(self, Node::Root(_))
+    }
+
+    /// Returns this node's symbol, if it has one.  (_Pop symbol_, _pop scoped symbol_, _push
+    /// symbol_, and _push scoped symbol_ nodes have symbols.)
+    pub fn symbol(&self) -> Option<Handle<Symbol>> {
+        match self {
+            Node::DropScopes(_) => None,
+            Node::ExportedScope(_) => None,
+            Node::InternalScope(_) => None,
+            Node::JumpTo(_) => None,
+            Node::PushScopedSymbol(node) => Some(node.symbol),
+            Node::PushSymbol(node) => Some(node.symbol),
+            Node::PopScopedSymbol(node) => Some(node.symbol),
+            Node::PopSymbol(node) => Some(node.symbol),
+            Node::Root(_) => None,
+            Node::Unknown(_) => None,
+        }
```

Comment:
Is there a requirement in Rust to enumerate all variants? Or could we collapse the Node types without a `Symbol` to a single `_ => None` match?

Reply:
No requirement, and I like the reduced clutter with a catch-all like you suggest.  :heavy_check_mark: